pyOCD/tools/pyocd.py

```diff
@@ -69,207 +76,490 @@ def dumpHexData(data, startAddress=0, width=8):
                     break
         print
 
-
 class ToolError(Exception):
     pass
 
+def cmdoptions(opts):
+    def process_opts(fn):
+        parser = optparse.OptionParser(add_help_option=False)
+        for opt in opts:
+            parser.add_option(opt)
+        def foo(inst, args):
+            namespace, other_args = parser.parse_args(args)
+            return fn(inst, namespace, other_args)
+        return foo
+    return process_opts
+
+class PyOCDConsole(object):
+    PROMPT = '>>> '
+
+    def __init__(self, tool):
+        self.tool = tool
+        self.last_command = ''
+
+    def run(self):
+        try:
+            while True:
+                try:
+                    line = raw_input(self.PROMPT)
+                    line = line.strip()
+                    if line:
+                        self.process_command_line(line)
+                        self.last_command = line
+                    elif self.last_command:
+                        self.process_command(self.last_command)
+                except KeyboardInterrupt:
+                    print
+        except EOFError:
+            print
+
+    def process_command_line(self, line):
+        for cmd in line.split(';'):
+            self.process_command(cmd)
+
+    def process_command(self, cmd):
+        try:
+            args = cmd.split()
+            cmd = args[0].lower()
+            args = args[1:]
+
+            # Handle help.
+            if cmd in ['?', 'help']:
+                self.show_help(args)
+                return
+
+            # Handle register name as command.
+            if cmd in pyOCD.target.cortex_m.CORE_REGISTER:
+                self.tool.handle_reg([cmd])
+                return
+
+            # Check for valid command.
+            if cmd not in self.tool.command_list:
+                print "Error: unrecognized command '%s'" % cmd
+                return
+
+            # Run command.
+            handler = self.tool.command_list[cmd]
+            handler(args)
+        except ValueError:
+            print "Error: invalid argument"
+            traceback.print_exc()
+        except pyOCD.transport.transport.TransferError:
+            print "Error: transfer failed"
+        except ToolError, e:
```

Comment:
optional - pep8 recommends using 'as' rather than a comma.  This would be "except ToolError as e:".